### ISSUE TITLE
multiple handler routing / "Next" responder

### ISSUE THREAD

**Issue by ramiroaisen** [actix/actix-web#1425]
It's there a way to pass a request to the next handler?
Something like
```rust
async fn handler() -> impl Responder {
  if some_condition() {
    Response::Ok()
  } else {
    Next
  }
}
```

**Comment by robjtede** [comment]
What would the “next handler” mean exactly?

**Comment by ramiroaisen** [comment]
> What would the “next handler” mean exactly?

It would mean that the request continues the handling process as if it did not match the handler
something like the ```(req, res, next <--)``` function in express (nodejs) if you know waht i mean.

**Comment by robjtede** [comment]
Alright, just checking there was a distinction here against the other semantic in Node land for `next` (which, in my experience with Koa, was yielding control to inner middlewares).

There's been a couple discussion in various issues that have mentioned having multiple handlers be able to answer the same query. Some of the discussions have been around guard matching but I like the idea of the express-like system where you can use "fallback handlers".

Just needs to pass request to the next out of the list of other, resource-matching handlers if the primary handler declares, by way of returning Next, that it isn't able to handle the current request. (It appears that current implementation will only service the last defined handler for a resource path but its plausible to change that.)

I see a couple issues that need to be addressed for any kind of multi-handler system to be considered viable:

- What happens when the last or only handler of a route returns Next, should it be treated as a 404, a 400, maybe a 500? The semantics of what Next means for every possibility should be clearly defineable and potentially recoverable/configurable with app data.
- How should ownership be handled? Obviously Rust's mutability and ownership rules are much tighter than Node's. It would be very easy to end up in situations where parts of the request are consumed or modified in the first handler but are expected to work in the second handler, with the full request. I can think of a few approaches for handling this but all have drawbacks for now.
- Should there be an alternative method to .to() for multiple handlers to prevent breaking existing routing tables that consider only the last entry for a path?

Interested to hear your thoughts @ramiroaisen and even @actix/contributors?

**Comment by ramiroaisen** [comment]
the next solution also brings the posibility to match two different string formats in the same slot of the path, the handler then check if the format is the required and if not just pass it to the next handler

> What happens when the last or only handler of a route returns Next, should it be treated as a 404, a 400, maybe a 500? The semantics of what Next means for every possibility should be clearly defineable and potentially recoverable/configurable with app data.

I think the better solution if there's no more handlers if just return an empty 404,

> How should ownership be handled? Obviously Rust's mutability and ownership rules are much tighter than Node's. It would be very easy to end up in situations where parts of the request are consumed or modified in the first handler but are expected to work in the second handler, with the full request. I can think of a few approaches for handling this but all have drawbacks for now.

I think it can be done if the handler returns ownership to the system
something like this:
```rust
async fn handler(request: Request) -> impl Responder {
  if some_condition() {
    Response::Ok()
  } else {
    Next(request)
  }
}
```

**Comment by robjtede** [comment]
Interested to get @actix/contributors take on this?

**Comment by shakyShane** [comment]
I'm also seeing a need for this.

My specific use-case is wanting to stack services 

```rust
.service(Files::new("/", "dir1"))
.service(Files::new("/", "dir2"))
.service(Proxy::new("https://example.com"))
```

Where the first 2 handlers would `next` if they cannot serve a specific file from disk (instead of erroring/404'in) finally ending up with a proxy services taking over the request.

**Comment by fakeshadow** [comment]
This is not useful or easy to do with built in services types of actix-web. 
The service chain in actix-web always want a `ServiceRequest` in and `ServiceResponse` out. Adding addional types to input or output would be major break change and additional overhead for everyone to pay for no matter they use it or not.

You can always build your own service types for nested service logic. Like `Files<Files<Proxy>>` would easily achieve what you want and it acts very similar to middlewares anyway(The difference would be with custom nested services you can have self defined input output type). 

TLDR: The built in service types of actix-web is to satisfy majority usage with a stable pattern and you can always customize services easily for your advanced usage.

**Comment by ModProg** [comment]
I have something somewhat related, but I could do it if I could access a Database in app_data from a guard.

**Comment by imgurbot12** [comment]
Reviving this thread from the dead.

I have a use-case where services are dynamically created in sequence based on a configuration file. Should the first service fail to find a result, it moves onto the next service until all are exhausted.

I cannot seemingly compose a static type definition wrapping the associated objects like `Files<Files<Proxy>>`. [`actix-files` has `default_handler`](https://docs.rs/actix-files/0.6.6/actix_files/struct.Files.html#method.default_handler). Is there any way to dynamically chain together a series of services to run in sequence if the last doesn't complete successfully?

**Comment by imgurbot12** [comment]
Update, I found a very overly complicated way of doing it. You can [write an actix-service that calls other actix-services in sequence](https://github.com/imgurbot12/bob/blob/master/src/modules/service.rs). I took inspiration from [shakyShane's actix-multi](https://github.com/shakyShane/actix-multi/blob/main/src/service.rs) and [actix-web's actix-files](https://github.com/actix/actix-web/blob/master/actix-files/src/service.rs#L35).

It's not trivial to implement not only because of all the boilerplate, but also because you have to implement a means to re-use the same [`Payload::Stream`](https://docs.rs/actix-web/latest/actix_web/dev/enum.Payload.html) included in the [`ServiceRequest`](https://docs.rs/actix-web/latest/actix_web/dev/struct.ServiceRequest.html).

Hopefully that helps somebody, thanks :)

**Comment by imgurbot12** [comment]
Just published a crate with a service that supports dynamic assembly and routing of actix-services if anyone wants to check it out :)

https://crates.io/crates/actix-chain
